typed/remove_test.go

```diff
@@ -0,0 +1,340 @@
+package typed
+
+import (
+	"fmt"
+	"testing"
+
+	"sigs.k8s.io/structured-merge-diff/v3/fieldpath"
+	"sigs.k8s.io/structured-merge-diff/v3/value"
+)
+
+func TestRemoveDeduced(t *testing.T) {
+
+	var cases = []struct {
+		object YAMLObject
+		remove *fieldpath.Set
+		expect YAMLObject
+	}{
+		{
+			object: `{}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: `{}`,
+		},
+		{
+			object: `{"a": "value"}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: ``,
+		},
+		{
+			object: `{"a": "value", "b": "value"}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: `{"b": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": {}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: `{"b": {}}`,
+		},
+		{
+			object: `{"a": "value", "b": {"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			expect: `{"a": "value"}`,
+		},
+		{
+			// this is okay, because in a deduced schema `b` is handled as a map not a struct
+			object: `{"a": "value", "b":{"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			expect: `{"a": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": {"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "c")),
+			expect: `{"a": "value", "b": null}`,
```

Comment:
now it's not {} or empty (it cant' be empty), but null. I know we talked about it, but it's always confusing.

Reply:
Yeah, I think we should discuss the {} vs null thing and put it down so we know why we do it that way (or not)
Not sure myself anymore...

Reply:
Removing an item from a directory should not "remove that directory" if it ends-up being empty. I'm sort of following the semantics of `rm -r` in my mind, and that seems wrong.